### ISSUE TITLE
NLB-IP for the pods that are assigned custom security groups

### ISSUE THREAD

**Issue by thapakazi** [kubernetes-sigs/aws-load-balancer-controller#2054]
***Query***
tldr;

`Is nlb-ip supported for pods with security groups (crd/securitygrouppolicy) ? 
`
my context: controller version 2.2.0, eks/k8s: "v1.19.8-eks-96780e",
I am trying use nlb for my service, that points to pods that are assigned with custom security groups.

The docs: https://docs.aws.amazon.com/eks/latest/userguide/network-load-balancing.html#network-load-balancing-service-sample-manifest
covers all requirements and assumptions. But I am unable to find my usecase*.

***Description***

my usecase*
1. I have pods that are assigned with custom security groups
2. service points to those pods and is configured as per the doc
```yaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    meta.helm.sh/release-name: myapp-backend
    meta.helm.sh/release-namespace: test
    service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: ip
    service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
    service.beta.kubernetes.io/aws-load-balancer-type: external
    #service.beta.kubernetes.io/aws-load-balancer-proxy-protocol: '*'
    #service.beta.kubernetes.io/aws-load-balancer-backend-protocol: http
    #service.beta.kubernetes.io/aws-load-balancer-ssl-ports: https
  labels:
    app.kubernetes.io/instance: myapp-backend
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: myapp-backend
    app.kubernetes.io/version: 1.16.0
    helm.sh/chart: myapp-backend-0.1.0
  name: myapp-backend
  namespace: test
spec:
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 8000
  - name: https
    port: 443
    protocol: TCP
    targetPort: 8000
  selector:
    app.kubernetes.io/instance: myapp-backend
    app.kubernetes.io/name: myapp-backend
  sessionAffinity: None
  type: LoadBalancer 

```

3. The nlb that is created has:
  - required subnets association and target groups
  - **but it fails to add the pod ip to nlb targets as I am not using Node Security Group, but rather bunch of other security group specific to my pod with custom CRD/ SecurityGroupPolicy**
  
  I found this doc:  https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/guide/service/nlb/#security-group
  
![screenshot-20210606-19:55:34](https://user-images.githubusercontent.com/1695906/120927612-30a5d080-c701-11eb-9844-b242452dd8a0.png)

but can't make much sense of it. 
```
In case of multiple security groups, the controller expects to find only one security group tagged with the Kubernetes cluster id. Controller will update the ingress rules on the security groups as per the service spec.
```

How would it identify security group that I attached to pod and make those adjustments ? 

I tried removing the security group for the pod and the above process: pointing target-groups -> service {pods } works as I would be using node security group in that case. 
  
 I wanted to do end-to-end ssl termination for my pods, but I now I am lost in some random rabbit hole.

**Comment by thapakazi** [comment]
Update: I tried to limit the security assigned to my pod to one, and it seems working. 

The nlb target is forwarding traffic to my pod, it also updated the sg rule on that custom sg assigned to the pod.

![screenshot-20210606-20:32:48](https://user-images.githubusercontent.com/1695906/120928905-584b6780-c706-11eb-9305-d0b6156622fa.png)
I had to manually add outbound rule though :( 

Now the question is how do I select the security group, in case I have multiple attached.

```
In case of multiple security groups, the controller expects to find only one security group tagged with the Kubernetes cluster id.
```
This is what I found on doc, but its not clear about tag, **what is key name and what is the value**

**Comment by thapakazi** [comment]
and little bit of digging, i think I found my answers here: https://github.com/kubernetes-sigs/aws-load-balancer-controller/blob/main/pkg/targetgroupbinding/networking_manager.go#L412-L438


tagging used is common tag: https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.2/deploy/subnet_discovery/#common-tag

```
key                                     value
kubernetes.io/cluster/${cluster-name}	owned or shared
```

I think documentation could be improved so next person doesn't get lost on this :grin:

**Comment by kishorj** [comment]
@thapakazi, thanks for sharing the details. Feel free to open a PR to update the documentation.
